# **HW 3**
### Josh Winnes
### Topics in Data Science
### September 5, 2024

## **Data Setup**

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from mlxtend.preprocessing import TransactionEncoder
from mlxtend.preprocessing import *
from mlxtend.frequent_patterns import association_rules
from mlxtend.frequent_patterns import fpgrowth
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import fpmax

import os
import sys

# Specify the path to the directory containing your script
script_directory = '/Users/joshwinnes/Library/Mobile Documents/com~apple~CloudDocs/Wheaton College/fall 2024/topics in data science/Math485/packages'
sys.path.append(script_directory)

# Now you can import your script
import dhs_util
from dhs_util import * 

os.chdir('/Users/joshwinnes/Library/Mobile Documents/com~apple~CloudDocs/Wheaton College/fall 2024/topics in data science/data')
df = pd.read_csv('dhs_service_records_synthesized_final.csv')

df = dhs_preprocessing(df)
df, service_map = add_service_label(df)
df = add_age_bin(df)

recipient = get_recipient_attribute(df)

/Users/joshwinnes/Library/Mobile Documents/com~apple~CloudDocs/Wheaton College/fall 2024/topics in data science/Math485/packages/dhs_util.py:42: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  recipient = df.groupby(df['id'].cat.remove_unused_categories()).agg(


## **Homework**
**Specifications**
- from the DHS synthetic dataset, prepare an analysis file for mining association rules about services recipients use, run association analysis using mlxtend package

frequent itemsets
- extract service itemsets that have support greater than 0.01using apriori
- extract service itemsets that have support greater than 0.0001 using fpgrowth
- extract service itemsets that have support greater than 0.0001 using hmine
- make observations of the difference between them
- extract service itemsets that have support greater than 0.0001 using fpmax
- make observations between fpmax and others

frequent itemsets with length (size of the itemsets)
- extract length-1 service itemsets that have support greater than 0.01 using apriori
- extract length-2 service itemsets that have support greater than 0.005 using hmine
- extract length-5 service itemsets that have support greater than 0.0001using fpmax

association rules
- extract association rules that have lift greater than 0.5, using itemsets from fpgrowth with support greater than 0.05
- extract association rules that have support greater than 0.0003, using itemsets from fpmax with support greater than 0.0001

make prediction (draw conclusion)
- if you were to make suggestions for DHS to consider increase the offering of one service, say, “Families_Receiving_Child_Welfare_Services” (S06), what other services would you suggest to offer together with? 
- run predict({“S06”}, rules, consequents_only=False), with rules generated from different thresholds and algorithms

**Deliverables**
- self contained python jupyter notebook that implements the specifications. 
- the notebook should not include “irrelevant” code chunks, it may include code chunks outside what is in the specification but you consider relevant that the reader (yourself and teacher) may be interested in reading. 


### **Exercise 1**
frequent itemsets
- extract service itemsets that have support greater than 0.01 using apriori

In [3]:
def oneHotCoding(serv_data):
    te = TransactionEncoder()
    te_ary = te.fit(serv_data).transform(serv_data)
    te_df = pd.DataFrame(te_ary, columns=te.columns_)
    return te_df


serv_data = []
for groups in df.groupby('id').groups.values():
    serv_data.append(df.loc[groups]['serv'].tolist())
serv_data[:10]
serv_oneHot = oneHotCoding(serv_data)

/var/folders/99/6591xgpn4sq1ypfgnrzy_7n80000gn/T/ipykernel_21240/2138446625.py:9: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  for groups in df.groupby('id').groups.values():


In [4]:
apriori(serv_oneHot, min_support=0.01) 

support         itemsets
0   0.013687              (2)
1   0.012649              (4)
2   0.153844              (8)
3   0.040882             (10)
4   0.941422             (11)
5   0.011915             (12)
6   0.103528             (13)
7   0.024002             (14)
8   0.012967             (16)
9   0.013573             (17)
10  0.031396             (18)
11  0.013260             (20)
12  0.012284          (2, 11)
13  0.018307          (8, 10)
14  0.139131          (8, 11)
15  0.103528          (8, 13)
16  0.019431          (8, 14)
17  0.013468          (8, 17)
18  0.032106         (10, 11)
19  0.011448         (11, 12)
20  0.094436         (11, 13)
21  0.022561         (11, 14)
22  0.010384         (17, 11)
23  0.013460         (17, 13)
24  0.016508      (8, 10, 11)
25  0.094436      (8, 11, 13)
26  0.018280      (8, 11, 14)
27  0.010285      (8, 17, 11)
28  0.013460      (8, 17, 13)
29  0.010277     (17, 11, 13)
30  0.010277  (8, 17, 11, 13)

- extract service itemsets that have support greater than 0.0001 using fpgrowth

In [5]:
freq_itemsets_fpgrowth = fpgrowth(serv_oneHot,min_support=0.0001,use_colnames=True)
freq_itemsets_fpgrowth

support                        itemsets
0    0.941422                           (S12)
1    0.153844                           (S09)
2    0.040882                           (S11)
3    0.103528                           (S14)
4    0.013687                           (S03)
..        ...                             ...
930  0.000111       (S10, S03, S14, S02, S12)
931  0.000111  (S10, S03, S09, S14, S02, S12)
932  0.000124                      (S12, S22)
933  0.000305                      (S01, S12)
934  0.000199                      (S12, S08)

[935 rows x 2 columns]

- extract service itemsets that have support greater than 0.0001 using hmine

In [ ]:
## Need help with this one

- make observations of the difference between them

*I'll be able to do this once I know how to do the one above*

- extract service itemsets that have support greater than 0.0001 using fpmax

In [6]:
freq_itemsets_fpmax = fpmax(serv_oneHot, min_support = 0.0001, use_colnames = True)
freq_itemsets_fpmax

support                        itemsets
0   0.000199                      (S12, S08)
1   0.000305                      (S01, S12)
2   0.000124                      (S12, S22)
3   0.000133            (S02, S12, S03, S04)
4   0.000131       (S03, S09, S14, S04, S12)
..       ...                             ...
59  0.000122  (S18, S11, S09, S21, S14, S12)
60  0.000210  (S18, S09, S21, S14, S15, S12)
61  0.000204       (S18, S19, S09, S14, S12)
62  0.000169       (S19, S09, S14, S15, S12)
63  0.000247       (S11, S19, S09, S14, S12)

[64 rows x 2 columns]

- make observations between fpmax and others

`fpmax` was much more selective than `fpgrowth`, even though they have the same minimum support, `fpmax` only gave me 63 itemsets while `fpgrowth` extracted 934 itemsets. Furthermore, `fpgrowth` gave itemsets of size 1 while `fpmax` did not. 

### **Exercise 2**

frequent itemsets with length (size of the itemsets)
- extract length-1 service itemsets that have support greater than 0.01 using apriori

In [8]:
itemsets_length2 = apriori(serv_oneHot, min_support=0.01)
itemsets_length2 = itemsets_length2[itemsets_length2['itemsets'].apply(lambda x: len(x)==2)]
itemsets_length2 

support  itemsets
12  0.012284   (2, 11)
13  0.018307   (8, 10)
14  0.139131   (8, 11)
15  0.103528   (8, 13)
16  0.019431   (8, 14)
17  0.013468   (8, 17)
18  0.032106  (10, 11)
19  0.011448  (11, 12)
20  0.094436  (11, 13)
21  0.022561  (11, 14)
22  0.010384  (17, 11)
23  0.013460  (17, 13)

- extract length-2 service itemsets that have support greater than 0.005 using hmine

- extract length-5 service itemsets that have support greater than 0.0001using fpmax

In [10]:
itemsets_length2_fpmax = fpmax(serv_oneHot, min_support = 0.0001)
itemsets_length2_fpmax = itemsets_length2_fpmax[itemsets_length2_fpmax['itemsets'].apply(lambda x: len(x) ==5)]
itemsets_length2_fpmax

support             itemsets
4   0.000131    (2, 3, 8, 11, 13)
18  0.000163   (8, 9, 11, 13, 17)
19  0.000157    (5, 8, 9, 11, 20)
21  0.000124   (8, 9, 11, 14, 20)
22  0.000208   (8, 9, 11, 13, 14)
23  0.000139   (8, 9, 11, 13, 20)
24  0.000197   (8, 9, 10, 11, 13)
25  0.000103   (5, 8, 13, 15, 20)
31  0.000103   (6, 8, 10, 11, 18)
33  0.000120    (2, 6, 8, 10, 11)
37  0.000131  (8, 13, 15, 17, 18)
38  0.000142  (8, 11, 13, 15, 18)
45  0.000120   (2, 8, 11, 12, 13)
46  0.000111  (8, 11, 12, 13, 16)
47  0.000150  (8, 11, 12, 13, 14)
49  0.000150  (8, 10, 11, 12, 13)
50  0.000111   (4, 8, 11, 13, 17)
53  0.000109  (8, 10, 11, 16, 20)
61  0.000204  (8, 11, 13, 17, 18)
62  0.000169  (8, 11, 13, 14, 18)
63  0.000247  (8, 10, 11, 13, 18)

### **Exercise 3**

association rules
- extract association rules that have lift greater than 0.5, using itemsets from fpgrowth with support greater than 0.05

In [13]:
def serv_rules(freq_itemsets,metrics,threshold):
    asso_rules = association_rules(freq_itemsets, metric=metrics, min_threshold=threshold)
    return asso_rules.sort_values(by='lift', ascending=False)[['antecedents', 'consequents', 'support', 'confidence', 'lift']]

In [15]:
itemsets_great_5 = fpgrowth(serv_oneHot, min_support=0.05)
rules_fpgrowth = serv_rules(itemsets_great_5,"lift",0.5)
rules_fpgrowth

antecedents consequents   support  confidence      lift
6      (8, 11)        (13)  0.094436    0.678758  6.556292
11        (13)     (8, 11)  0.094436    0.912184  6.556292
2          (8)        (13)  0.103528    0.672938  6.500073
3         (13)         (8)  0.103528    1.000000  6.500073
8     (11, 13)         (8)  0.094436    1.000000  6.500073
9          (8)    (11, 13)  0.094436    0.613843  6.500073
4         (11)        (13)  0.094436    0.100312  0.968942
5         (13)        (11)  0.094436    0.912184  0.968942
7      (8, 13)        (11)  0.094436    0.912184  0.968942
10        (11)     (8, 13)  0.094436    0.100312  0.968942
0          (8)        (11)  0.139131    0.904362  0.960634
1         (11)         (8)  0.139131    0.147788  0.960634

- extract association rules that have support greater than 0.0003, using itemsets from fpmax with support greater than 0.0001

In [21]:
itemsets_great_0001 = fpmax(serv_oneHot, min_support = 0.0001)
rules_fpmax = association_rules(itemsets_great_0001, 'support', 0.0003, support_only = True)
rules_fpmax

antecedents          consequents  antecedent support  \
0                   (0)                 (11)                 NaN   
1                  (11)                  (0)                 NaN   
2   (8, 11, 13, 14, 16)                 (17)                 NaN   
3   (8, 11, 13, 14, 17)                 (16)                 NaN   
4   (8, 11, 13, 16, 17)                 (14)                 NaN   
..                  ...                  ...                 ...   
59                 (11)  (8, 13, 14, 16, 17)                 NaN   
60                 (13)  (8, 11, 14, 16, 17)                 NaN   
61                 (14)  (8, 11, 13, 16, 17)                 NaN   
62                 (16)  (8, 11, 13, 14, 17)                 NaN   
63                 (17)  (8, 11, 13, 14, 16)                 NaN   

    consequent support   support  confidence  lift  leverage  conviction  \
0                  NaN  0.000305         NaN   NaN       NaN         NaN   
1                  NaN  0.000305         NaN   NaN       NaN         NaN   
2                  NaN  0.000326         NaN   NaN       NaN         NaN   
3                  NaN  0.000326         NaN   NaN       NaN         NaN   
4                  NaN  0.000326         NaN   NaN       NaN         NaN   
..                 ...       ...         ...   ...       ...         ...   
59                 NaN  0.000326         NaN   NaN       NaN         NaN   
60                 NaN  0.000326         NaN   NaN       NaN         NaN   
61                 NaN  0.000326         NaN   NaN       NaN         NaN   
62                 NaN  0.000326         NaN   NaN       NaN         NaN   
63                 NaN  0.000326         NaN   NaN       NaN         NaN   

    zhangs_metric  
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN  
..            ...  
59            NaN  
60            NaN  
61            NaN  
62            NaN  
63            NaN  

[64 rows x 10 columns]

### **Exercise 3**

make prediction (draw conclusion)
- if you were to make suggestions for DHS to consider increase the offering of one service, say, “Families_Receiving_Child_Welfare_Services” (S06), what other services would you suggest to offer together with? 
- run predict({“S06”}, rules, consequents_only=False), with rules generated from different thresholds and algorithms


In [22]:
def predict(antecedent, rules, consequents_only = False):
    
    # get the rules for this antecedent
    preds = rules[rules['antecedents'] == antecedent]
    if consequents_only:
        # a way to convert a frozen set with one element to string
        preds = preds['consequents'].apply(iter).apply(next)
    return preds

In [24]:
itemsets_fpgrowth1 = fpgrowth(serv_oneHot, min_support=0.05)
itemsets_fpgrowth2 = fpgrowth(serv_oneHot, min_support = 0.01)
itemsets_fpgrowth3 = fpgrowth(serv_oneHot, min_support = 0.1)
itemsets_fpgrowth4 = fpgrowth(serv_oneHot, min_support = 0.25)

itemsets_fpmax1 = fpmax(serv_oneHot, min_support = 0.001)
itemsets_fpmax2 = fpmax(serv_oneHot, min_support = 0.005)
itemsets_fpmax3 = fpmax(serv_oneHot, min_support = 0.0001)
itemsets_fpmax4 = fpmax(serv_oneHot, min_support = 0.0025)

itemsets_fpgrowth = [itemsets_fpgrowth1, itemsets_fpgrowth2, itemsets_fpgrowth3, itemsets_fpgrowth4]
itemsets_fpmax = [itemsets_fpmax1, itemsets_fpmax2, itemsets_fpmax3, itemsets_fpmax4]

In [37]:
predictions = []
for itemset in itemsets_fpgrowth:
    rules = serv_rules(itemset, 'confidence', 0.5)
    prediction = predict({'S06'}, rules, consequents_only = False)
    predictions.append(prediction)
    